sockeye/inference.py

```diff
@@ -1441,6 +1482,12 @@ def _beam_search(self,
             # finished rows are inf everywhere except column zero, which holds the accumulated model score
             scores = self._update_scores.forward(scores, finished, inactive, scores_accumulated, self.inf_array, pad_dist)
 
+            # Mark entries that should be blocked as having a score of np.inf
+            if self.global_avoid_trie or any(raw_avoid_list):
+                indices = avoid_states.avoid()
+                if any(indices):
+                    scores[indices] = np.inf
```

Comment:
this is much clearer thanks. 
At some point (not scope of this PR of course) we should think about how to move the constrained/avoid logic to proper operators/gluon blocks that are potentially hybridizable. As seen recently, these imperative indexing ops are quite slow.